pkg/controller/podautoscaler/replica_calculator_test.go

```diff
@@ -0,0 +1,546 @@
+			targetUtilization:   15.0,
+			expectedUtilization: 15.0,
+		},
+	}
+	tc.runTest(t)
+}
+
+func TestReplicaCalcScaleDown(t *testing.T) {
+	tc := replicaCalcTestCase{
+		currentReplicas:  5,
+		expectedReplicas: 3,
+		resource: &resourceInfo{
+			name:     api.ResourceCPU,
+			requests: []resource.Quantity{resource.MustParse("1.0"), resource.MustParse("1.0"), resource.MustParse("1.0"), resource.MustParse("1.0"), resource.MustParse("1.0")},
+			levels:   []int64{100, 300, 500, 250, 250},
+
+			targetUtilization:   50,
+			expectedUtilization: 28,
+		},
+	}
+	tc.runTest(t)
+}
+
+func TestReplicaCalcScaleDownCM(t *testing.T) {
+	tc := replicaCalcTestCase{
+		currentReplicas:  5,
+		expectedReplicas: 3,
+		metric: &metricInfo{
+			name:                "qps",
+			levels:              []float64{12.0, 12.0, 12.0, 12.0, 12.0},
+			targetUtilization:   20.0,
+			expectedUtilization: 12.0,
+		},
+	}
+	tc.runTest(t)
+}
+
+func TestReplicaCalcScaleDownIgnoresUnreadyPods(t *testing.T) {
+	tc := replicaCalcTestCase{
+		currentReplicas:  5,
+		expectedReplicas: 2,
+		podReadiness:     []api.ConditionStatus{api.ConditionTrue, api.ConditionTrue, api.ConditionTrue, api.ConditionFalse, api.ConditionFalse},
+		resource: &resourceInfo{
+			name:     api.ResourceCPU,
+			requests: []resource.Quantity{resource.MustParse("1.0"), resource.MustParse("1.0"), resource.MustParse("1.0"), resource.MustParse("1.0"), resource.MustParse("1.0")},
+			levels:   []int64{100, 300, 500, 250, 250},
+
+			targetUtilization:   50,
+			expectedUtilization: 30,
+		},
+	}
+	tc.runTest(t)
+}
+
+func TestReplicaCalcTolerance(t *testing.T) {
+	tc := replicaCalcTestCase{
+		currentReplicas:  3,
+		expectedReplicas: 3,
+		resource: &resourceInfo{
+			name:     api.ResourceCPU,
+			requests: []resource.Quantity{resource.MustParse("0.9"), resource.MustParse("1.0"), resource.MustParse("1.1")},
+			levels:   []int64{1010, 1030, 1020},
+
+			targetUtilization:   100,
+			expectedUtilization: 102,
+		},
+	}
+	tc.runTest(t)
+}
+
+func TestReplicaCalcToleranceCM(t *testing.T) {
+	tc := replicaCalcTestCase{
+		currentReplicas:  3,
+		expectedReplicas: 3,
+		metric: &metricInfo{
+			name:                "qps",
+			levels:              []float64{20.0, 21.0, 21.0},
+			targetUtilization:   20.0,
+			expectedUtilization: 20.66666,
+		},
+	}
+	tc.runTest(t)
+}
+
+func TestReplicaCalcSuperfluousMetrics(t *testing.T) {
+	tc := replicaCalcTestCase{
+		currentReplicas:  4,
+		expectedReplicas: 24,
+		resource: &resourceInfo{
+			name:                api.ResourceCPU,
+			requests:            []resource.Quantity{resource.MustParse("1.0"), resource.MustParse("1.0"), resource.MustParse("1.0"), resource.MustParse("1.0")},
+			levels:              []int64{4000, 9500, 3000, 7000, 3200, 2000},
+			targetUtilization:   100,
+			expectedUtilization: 587,
+		},
+	}
+	tc.runTest(t)
+}
+
+func TestReplicaCalcMissingMetrics(t *testing.T) {
+	tc := replicaCalcTestCase{
+		currentReplicas:  4,
+		expectedReplicas: 3,
```

Comment:
We still return the original, undoctored utilization, as discussed way earlier in this thread